src/coreclr/src/tools/Common/Sorting/MergeSort.cs

```diff
@@ -0,0 +1,185 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using ILCompiler.Sorting.Implementation;
+
+namespace ILCompiler
+{
+    public static class MergeSortApi
+    {
+        private const bool UseStandardSort = false; // Set this to true to use List<T>.Sort and Array.Sort instead of the mergesort algorithm
+                                                    // Used to measure the performance impact of using this custom sort algorithm
+        internal const int ParallelSortThreshold = UseStandardSort ? Int32.MaxValue : 4000; // Number empirically measured. 
+
+        private static class SortAsEqualBehavior<T>
+        {
+            public static Action<T,T> FailSort = DoFailSort;
+
+            private static void DoFailSort(T t1, T t2)
+            {
+                throw new InvalidOperationException();
+            }
+        }
+
+        // Parallel sorting api which will sort in parallel when appropriate
+        public static void ParallelSort<T>(this List<T> listToSort, Comparison<T> comparison)
```

Comment:
Nit - could we just use ISortableDataStructureAccessor as the "listToSort" to avoid having to duplicate the method so many times?